src/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs

```diff
@@ -250,8 +259,41 @@ public String Name
         {
             get
             {
-                Contract.Ensures(Contract.Result<String>() != null);
-                return Path.GetFileName(FullName);
+                // The Full path should be parsed based on the file format of the zip's source OS,
+                // not this OS. Because of this we can only use Path.GetFileName if the source OS is
+                // equal to this OS. Otherwise we resort to manual parsing.
+                string path = FullName;
+                Debug.Assert(path != null);
+                if (path.IndexOfAny(Path.GetInvalidPathChars()) >= 0)  // Validation checking is done based on current OS, not source OS.
```

Comment:
`Path.GetInvalidPathChars()` always allocates and return a new array. Should `PathInternal.InvalidPathChars` under src/Common be used instead, to avoid this?